sw/airborne/firmwares/rotorcraft/oneloop/oneloop_andi.c

```diff
@@ -0,0 +1,1524 @@
+  att_1l[2] = eulers_zxy.psi                        * use_increment;
+  oneloop_andi_propagate_filters();   //needs to be after update of attitude vector
+  att_d[0]  = rates_filt_fo[0].last_out             * use_increment;
+  att_d[1]  = rates_filt_fo[1].last_out             * use_increment;
+  att_d[2]  = rates_filt_fo[2].last_out             * use_increment;
+  att_2d[0] = ang_acc[0]                            * use_increment;
+  att_2d[1] = ang_acc[1]                            * use_increment;
+  att_2d[2] = ang_acc[2]                            * use_increment;
+  // (2) Position related
+  pos_1l[0] = stateGetPositionNed_f()->x;   
+  pos_1l[1] = stateGetPositionNed_f()->y;   
+  pos_1l[2] = stateGetPositionNed_f()->z;   
+  pos_d[0]  = stateGetSpeedNed_f()->x;      
+  pos_d[1]  = stateGetSpeedNed_f()->y;      
+  pos_d[2]  = stateGetSpeedNed_f()->z;      
+  pos_2d[0] = lin_acc[0];
+  pos_2d[1] = lin_acc[1];
+  pos_2d[2] = lin_acc[2];
+  
+  // Update the effectiveness matrix used in WLS
+  int8_t i;
+  for (i = 0; i < ANDI_OUTPUTS; i++) {
+    bwls_1l[i] = g1g2_1l[i];
+  }
+
+  // Run the Reference Model (RM)
+  oneloop_andi_RM(half_loop, PSA_des, rm_order_h, rm_order_v);
+  // Generate pseudo control for stabilization vector (nu) based on error controller
+  if(half_loop){
+    nu[0] = 0.0;
+    nu[1] = 0.0;
+    nu[2] = a_thrust;
+    nu[3] = ec_3rd(att_ref[0], att_d_ref[0], att_2d_ref[0], att_3d_ref[0], att_1l[0], att_d[0], att_2d[0], k_phi_e, k_p_e, k_pdot_e);
+    nu[4] = ec_3rd(att_ref[1], att_d_ref[1], att_2d_ref[1], att_3d_ref[1], att_1l[1], att_d[1], att_2d[1], k_theta_e, k_q_e, k_qdot_e);
+    nu[5] = ec_2rd(att_d_ref[2], att_2d_ref[2], att_3d_ref[2], att_d[2], att_2d[2], k_r_e, k_r_d_e);
+  }else{
+    float k1_att_e[3] = {k_phi_e,  k_theta_e, k_psi_e};
+    float k2_att_e[3] = {k_p_e,    k_q_e,     k_psi_d_e};
+    float k3_att_e[3] = {k_pdot_e, k_qdot_e,  k_psi_2d_e};
+    float y_4d_att[3];
+    ec_3rd_att(y_4d_att, att_ref, att_d_ref, att_2d_ref, att_3d_ref, att_1l, att_d, att_2d, k1_att_e, k2_att_e, k3_att_e);
+    nu[0] = ec_3rd(pos_ref[0], pos_d_ref[0], pos_2d_ref[0], pos_3d_ref[0], pos_1l[0], pos_d[0], pos_2d[0], k_N_e, k_vN_e, k_aN_e);
+    nu[1] = ec_3rd(pos_ref[1], pos_d_ref[1], pos_2d_ref[1], pos_3d_ref[1], pos_1l[1], pos_d[1], pos_2d[1], k_E_e, k_vE_e, k_aE_e);
+    nu[2] = ec_3rd(pos_ref[2], pos_d_ref[2], pos_2d_ref[2], pos_3d_ref[2], pos_1l[2], pos_d[2], pos_2d[2], k_D_e, k_vD_e, k_aD_e); 
+    nu[3] = y_4d_att[0];  
+    nu[4] = y_4d_att[1]; 
+    nu[5] = y_4d_att[2]; 
+  }
+
+  // Calculate the min and max increments
+  for (i = 0; i < ANDI_NUM_ACT_TOT; i++) {
+    if(i<ANDI_NUM_ACT){
+      du_min_1l[i]  = (act_min[i] - use_increment * actuator_state_1l[i])/ratio_u_un[i];//
+      du_max_1l[i]  = (act_max[i] - use_increment * actuator_state_1l[i])/ratio_u_un[i];//
+      du_pref_1l[i] = (u_pref[i]  - use_increment * actuator_state_1l[i])/ratio_u_un[i];//
```

Comment:
Check for division by zero

Reply:
Solved by bounding the variable at generation